### PR TITLE
Beta 7

### PR DESCRIPTION
**Pull request by adrien-coye**
### Changes
- feat(UploadOperation): at restart, an invalid upload session will be regenerated before we go further.
- refactor(UploadOperation+Error): reviewed and reworked error handling for more coherency.
- feat(UploadOperation): delete remote uplaod session when cleaning session if it is still valid.
- chore: Update to latest package versions
- More breadcrumbs for app lifecycle and background task scheduling events
- feat(AppDelegate): Suspend upload and download queue on app termination (user intend to stop the app)
- fix(UploadOperation): retry count at zero will throw
- fix(UploadFile): the identifier is always a UUID.
- feat(UploadFile): New fields to track FileProviderIdentifier and PHAsset.localIdentifier
- chore(UploadOperation): Minor refactor of init method

### CODE REVIEW

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift [hunk 1]
```diff
@@ -422,6 +445,30 @@ public final class UploadOperation: AsynchronousOperation, UploadOperationable,
...
+            let driveId = file.driveId
+            let userId = file.userId
+            let uploadingSession = file.uploadingSession?.detached()
+            self.enqueue {
+                guard let session = uploadingSession,
+                      !session.isExpired else {
+                    return
+                }
+
+                guard let driveFileManager = try? self.getDriveFileManager(for: driveId, userId: userId) else {
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:458]: Do we really want to strongly retain self ?

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift [hunk 2]
```diff
@@ -422,6 +445,30 @@ public final class UploadOperation: AsynchronousOperation, UploadOperationable,
         SentryDebug.uploadOperationCleanSessionBreadcrumb(uploadFileId)
 
         let cleanFileClosure: (UploadFile) -> Void = { file in
+            // Clean the remote session, if valid. Invalid ones are already gone server side.
+            let driveId = file.driveId
+            let userId = file.userId
+            let uploadingSession = file.uploadingSession?.detached()
+            self.enqueue {
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:452]: Do we really want to strongly retain self ?

File: kDriveCore/Data/Models/Upload/UploadFile.swift [hunk 3]
```diff
@@ -60,6 +60,8 @@ public class UploadFile: Object, UploadFilable {
     @Persisted(primaryKey: true) public var id = ""
     @Persisted public var name = ""
     @Persisted var relativePath = ""
+    @Persisted var fileProviderItemIdentifier: String?  // NSFileProviderItemIdentifier if any
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/Models/Upload/UploadFile.swift:63]: A migration is needed
```
kDriveCore/Logging.swift:58: Fatal error: Failed creating realm Migration is required due to the following errors:
- Property 'UploadFile.fileProviderItemIdentifier' has been added.
- Property 'UploadFile.assetLocalIdentifier' has been added.
```

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift (hunk 2 above)
Comment (adrien-coye) [reply kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:452]: Every use of `enqueue` `enqueueCatching` is making a strong reference on self in master.  (I changed that in the experimental upload from memory branch without much success solving my actual leaks _on this experimental branch_)

We could switch to [weak self] everywhere if you want. I do not see a reason to do so as I do not see retain cycles. (the UploadOperation is released when the upload is complete)

Should I use [weak self] everywhere in `enqueue` `enqueueCatching` you think ?

File: kDriveCore/Data/Models/Upload/UploadFile.swift (hunk 3 above)
Comment (adrien-coye) [reply kDriveCore/Data/Models/Upload/UploadFile.swift:63]: Interesting, a simple add only of a field needs a migration, unlike `Mail` ?

Comment (adrien-coye) [reply kDriveCore/Data/Models/Upload/UploadFile.swift]: Did the migration, managed to keep track of the PHAsset id during the migration when possible.
Tested a migration with both normal uploaded files and photo sync, all good.

File: kDriveCore/Data/Cache/DriveFileManager.swift [hunk 4]
```diff
@@ -85,6 +85,31 @@ public final class DriveFileManager {
...
+                        }
+
+                        let type: String? = oldObject?["rawType"] as? String ?? nil
+
+                        switch type {
+                        case UploadFileType.phAsset.rawValue:
+                            // The object was from a phAsset source, the id has to be the `LocalIdentifier`
+                            let oldAssetIdentifier: String? = oldObject?["id"] as? String ?? nil
+
+                            newObject["assetLocalIdentifier"] = oldAssetIdentifier
```
Comment (adrien-coye) [inline kDriveCore/Data/Cache/DriveFileManager.swift:103]: TODO `newObject["id"] = UUID().string…`

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**